feat(preprocessing): Generate annotations file in the nextclade preprocessing pipeline

fhennig opened the pull request:
resolves #4030 

based on #4076 (Anyas initial work on this) and #4344 (previous PR with feedback in it)

This PR adds support for processing annotations as EMBL files in the Nextclade preprocessing pipeline, and then uploads these files to the backend. Every nextclade-processed sequence can now get this annotation file, which we can then in the future use to submit to ENA.

### Tests
- test cases added with test files for regression. One example west nile, one example cchf
- integration test: A new test that checks on a submission of the annotations file is there.

### Screenshot


https://github.com/user-attachments/assets/d4942ba7-73ad-425c-95b7-484a9ecd115f

### How to enable this new feature on PPX?

```
s3:
  enabled: true
...
defaultOrganismConfig: &defaultOrganismConfig
  schema: &schema
...
    files:
      - name: annotations
...
  preprocessing:
    - &preprocessing
      image: ghcr.io/loculus-project/preprocessing-nextclade
...
      configFile: &preprocessingConfigFile
...
        create_embl_file: true
```



### PR Checklist
- [x] All necessary documentation has been adapted.
- [x] The implemented feature is covered by appropriate, automated tests.
- [x] Any manual testing that has been done is documented (i.e. what exactly was tested?)

🚀 Preview: https://generate-annotations-in-p.loculus.org

fhennig committed:
implementation

fhennig committed:
simplify fetching

anna-parker commented on preprocessing/nextclade/src/loculus_preprocessing/embl.py:
```diff
@@ -0,0 +1,261 @@
+    inclusive end (the default in nextclade is exclusive end)
+    """
+    # Map from nextclade attribute names to EMBL attribute names
+    attribute_map = {
```
Yes we need this it is important - this would be where we would add the mapping from `phase` to `codon_start` (https://github.com/loculus-project/loculus/pull/4076#issuecomment-2907878600) - I would add it and check the output embl flat files for differences

anna-parker replied:
This is important for the validity of the embl files IMO I think it should be verified before this is merged - I will try to get to that by this Friday

anna-parker replied:
done

corneliusroemer commented on integration-tests/tests/pages/search.page.ts:
```diff
@@ -75,9 +75,14 @@ export class SearchPage {
     }
 
-    async clickOnSequence(rowIndex = 0) {
+    async clickOnSequence(rowIndex = 0): Promise<string | null> {
```
```suggestion
    async clickOnSequenceAndGetAccession(rowIndex = 0): Promise<string | null> {
```

corneliusroemer commented on integration-tests/tests/pages/search.page.ts:
```diff
@@ -75,9 +75,14 @@ export class SearchPage {
-        await rows.nth(rowIndex).click();
+        const row = rows.nth(rowIndex);
+        const rowText = await row.innerText();
+        const loculusIdMatch = rowText.match(/LOC_[A-Z0-9]+\.[0-9]+/);
```
We usually call this `accessionVersion` not `loculusId`

corneliusroemer commented on integration-tests/tests/specs/features/annotations.dependent.spec.ts:
```diff
@@ -0,0 +1,90 @@
+    test.only('should have an embl file in the Files section', async ({ page }) => {
+        await searchPage.ebolaSudan();
+
+        const loculus_id = await searchPage.clickOnSequence(0);
```
How do you know the first sequence is going to remain the same one? That seems like a too strong assumption that will break as soon as:
- a newer sequence gets submitted
- or we change sort order (which shouldn't affect thsi test here)

fhennig replied:
i've changed the test to not use the pre-created sequence at all.

corneliusroemer commented on preprocessing/nextclade/src/loculus_preprocessing/backend.py:
```diff
@@ -119,8 +122,9 @@ def fetch_unprocessed_sequences(
     logger.info(
-        f"Unprocessed data from backend: status code {response.status_code}, "
-        "request id: {response.headers.get('x-request-id')}"
+        "Unprocessed data from backend: status code %s, request id: %s",
```
I don't understand why we should make this change?

fhennig commented on kubernetes/loculus/values.yaml:
```diff
@@ -63,6 +63,8 @@ defaultOrganismConfig: &defaultOrganismConfig
       externalFields:
         - ncbiReleaseDate
     richFastaHeaderFields: ["displayName"]
+    files:
```
Yes they are.

fhennig committed:
adress review

corneliusroemer commented:
Integration test fails: https://github.com/loculus-project/loculus/actions/runs/15845676604/job/44667172420#step:18:188

```
    + {"type":"about:blank","title":"Forbidden","status":403,"detail":"Authentication needed: the file is not public","instance":"/files/get/LOC_000001Y/1/annotations/LOC_000001Y.1.embl"}
```

fhennig commented:
I know, I can see it too, but thanks. Please engage with the discussion above.

fhennig committed:
Change test structure -> maybe fix?

fhennig committed:
add mapping 'phase' -> 'codon_start'

theosanderson commented on kubernetes/loculus/values.yaml:
```diff
@@ -1199,9 +1201,12 @@ defaultOrganismConfig: &defaultOrganismConfig
         - "prepro"
       replicas: 2
       configFile: &preprocessingConfigFile
+        scientific_name: "Sudan ebolavirus"
```
For now I would just go with what we have here. We could consider in future whether we want to bake in a concept of a "scientific name" to all preprocessing pipelines.

anna-parker replied:
I think @corneliusroemer just doesn't want the duplication because this could mean that we forgot to change a field in prepro while updating in ena deposition - generally having two fields to update can be challenging for maintainability and I think we should try to avoid that as well. But I am also fine with moving that work to a later PR

anna-parker commented on preprocessing/nextclade/src/loculus_preprocessing/config.py:
```diff
@@ -52,6 +72,12 @@ def load_config_from_yaml(config_file: str, config: Config | None = None) -> Con
+                for embl_key, embl_value in value.items():
+                    if hasattr(config.embl, embl_key) and embl_value is not None:
+                        setattr(config.embl, embl_key, embl_value)
+            else:
+                setattr(config, key, value)
```
You already do this above

```suggestion
```

anna-parker commented on preprocessing/nextclade/src/loculus_preprocessing/embl.py:
```diff
@@ -0,0 +1,262 @@
+
+
+def get_molecule_type(molecule_type: str) -> MoleculeType:
+    # Dummy enum for MoleculeType
```
```suggestion
```

anna-parker commented on preprocessing/nextclade/tests/test_data/west_nile_1/sequence.fa:
```diff
@@ -0,0 +1,2 @@
+>foo
```
```suggestion
>PQ483500.1
```

anna-parker commented on preprocessing/nextclade/src/loculus_preprocessing/embl.py:
```diff
@@ -0,0 +1,262 @@
+                for old_key, new_key in cds_attributes_map.items()
+                if old_key in attributes_cds
+            }
+            qualifiers["codon_start"] = 1
```
```suggestion
            # codon_start and phase define the offset at which the first 
# complete codon of a coding feature can be found, relative 
# to the first base of that feature. 
# Phase is 0-indexed, codon_start is 1 indexed
              qualifiers["codon_start"] = qualifiers.get("codon_start", 0) + 1
```

anna-parker commented on preprocessing/nextclade/tests/test_data/cchf_1/result.embl:
```diff
@@ -0,0 +1,529 @@
+FT                   /collection_date="1970"
+FT   gene            64..5118
+FT                   /product="GPC"
+FT                   /codon_start=1
```
```suggestion
```

anna-parker commented on preprocessing/nextclade/tests/test_data/cchf_1/result.embl:
```diff
@@ -0,0 +1,529 @@
+FT                   /collection_date="1970"
+FT   gene            237..1685
+FT                   /product="NP"
+FT                   /codon_start=1
```
```suggestion
```

anna-parker commented on integration-tests/tests/specs/features/annotations.spec.ts:
```diff
@@ -0,0 +1,127 @@
+FT   gene            1..945
+FT                   /gene="NP"
+FT                   /product="NP"
+FT                   /codon_start=1
```
```suggestion
```

anna-parker commented on preprocessing/nextclade/tests/test_data/cchf_1/result.embl:
```diff
@@ -0,0 +1,529 @@
+FT                   /collection_date="1970"
+FT   gene            72..11909
+FT                   /product="RdRp"
+FT                   /codon_start=1
```
```suggestion
```

anna-parker commented on integration-tests/tests/specs/features/annotations.spec.ts:
```diff
@@ -0,0 +1,127 @@
+import { test } from '../../fixtures/group.fixture';
+import { SingleSequenceSubmissionPage } from '../../pages/submission.page';
+
+test.describe('Sequence Preview Annotations', () => {
```
this test sadly seems quite flaky

anna-parker committed:
Anyas suggestions

anna-parker committed:
fix formatting

anna-parker committed:
await async function

anna-parker committed:
feat: fix integration tests

fhennig commented on integration-tests/tests/specs/features/annotations.spec.ts:
```diff
@@ -0,0 +1,127 @@
+test.describe('Sequence Preview Annotations', () => {
+    let searchPage: SearchPage;
+
+    test.beforeEach(async ({ page }) => {
```
I duplicated it in hopes of making the test less flaky, but it seems to not have worked I guess.

anna-parker committed:
fixup

anna-parker committed:
fixup

anna-parker commented on integration-tests/tests/pages/search.page.ts:
```diff
@@ -27,9 +27,12 @@ export class SearchPage {
     }
 
     async enableSearchFields(...fieldLabels: string[]) {
```
@theosanderson and @fhennig this subfunction is really flaky - it has been the cause of failing integration tests for most of this PR... I tried to now add another await but it doesn't seem to suffice - do you have any ideas? (When I run locally I see that the pop up is not opening although we navigate to the `'Add Search Fields'` and it looks like it is clicked)

anna-parker replied:
weirdly the tests only fail on chromium

theosanderson replied:
I mean what happens if one reverts the changes to this function? Or what are they needed for?

theosanderson replied:
It's not clear to me why we we need to wait for something to be visible, for example (that's normally inherent)

anna-parker reviewed (commented):
Hey this looks pretty good now! As soon as the integration tests pass and are deflaked I am happy to approve :-) 

Update: and we should merge in https://github.com/loculus-project/loculus/pull/4507

Update2: Just wanted to mention again - I initially asked for the prepro unit tests to just include a call to `nextclade`. The whole reason is to have a regression test that will error if a new `nextclade` version changes their `annotation.json` structure and that breaks flat file creation. I still think this is actually a better option than having an integration test. It will also mean we can remove the `annotation.json` files and just run `nextclade` on the fastas which is much easier (and will remove 1500 lines of code)

anna-parker committed:
lengthen timeout

anna-parker committed:
go back to previous state

anna-parker committed:
test if waiting for button helps

anna-parker committed:
trim Ns

fhennig committed:
simplify

fhennig committed:
slightly clean up test

corneliusroemer committed:
Fix lints

corneliusroemer commented:
Integration tests fail because revokation entry does not have annotations key in get-released-data crashing silo prepro.

corneliusroemer committed:
Fix expected embl content

corneliusroemer commented:
We might want to remove the empty line between sequences in the multi-segmented case:
<img width="725" alt="image" src="https://github.com/user-attachments/assets/1a4c4d99-a1f8-4c68-91f5-1a46a12a1686" />

The spec says it should be a single line with `//` to separate, it doesn't mention an empty line.

corneliusroemer committed:
fix: remove empty lines between segments

corneliusroemer committed:
fix test

corneliusroemer committed:
Deduplicate inherently coupled ena deposition related configs (#4507)

It doesn't make sense to duplicate the scientific name. It's:
- Annoying for maintainers
- Prone to cause bugs
- The annotation file is inherently coupled to ENA deposition, that's
the reason we're putting those annotations in there

Without this, we'd need to add 40 lines of yaml to our PPX production
yaml: 10 organisms x 2 processing versions x 2 lines per version

Co-authored-by: Chaoran Chen <mail@chaoran-chen.de>

anna-parker commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -881,6 +882,43 @@ class SubmissionDatabaseService(
     }
 
+    fun getCurrentPipelines(): List<String> {
+        // TODO DEBUG function, do not merge
```
this as well?

anna-parker replied:
outdated

anna-parker commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -3,6 +3,7 @@ package org.loculus.backend.service.submission
 import com.fasterxml.jackson.core.JacksonException
 import com.fasterxml.jackson.databind.ObjectMapper
 import com.fasterxml.jackson.module.kotlin.readValue
+import kotlinx.coroutines.selects.select
```
I don't think this is required?

anna-parker replied:
outdated

anna-parker commented on integration-tests/tests/pages/search.page.ts:
```diff
@@ -75,9 +78,14 @@ export class SearchPage {
     }
 
-    async clickOnSequence(rowIndex = 0) {
+    async clickOnSequenceAndGetAccession(rowIndex = 0): Promise<string | null> {
```
I would keep the `clickOnSequence` as there a lot of tests that use it and don't use the accession

fhennig replied:
haha ok I'll add a second function.

anna-parker replied:
done

anna-parker reviewed (commented):
Hey this is looking really good but would it be possible to put the recent backend changes into a separate PR? This PR is very large and the changes aren't linked to this but solve a bug in the file sharing feature - happy to approve once they are in their own PR :-)

fhennig commented on backend/src/main/kotlin/org/loculus/backend/service/submission/SubmissionDatabaseService.kt:
```diff
@@ -881,6 +882,43 @@ class SubmissionDatabaseService(
+            .map {
+                it[SequenceEntriesView.releasedAtTimestampColumn]
+            }
+            .first()
```
otherwise it's a list, not a single item.

anna-parker replied:
outdated

fhennig commented on integration-tests/tests/specs/features/annotations.dependent.spec.ts:
```diff
@@ -0,0 +1,94 @@
+        );
+
+        // use this to find our pre-made sequence
+        await searchPage.fill('Author affiliations', 'Patho Institute, Paris');
```
Hmm, I guess it makes it a bit more unique to add this, but it already works the way it is, doesn't it?

anna-parker committed:
dont await accession if not required for the test

anna-parker committed:
fix

anna-parker commented on integration-tests/tests/pages/search.page.ts:
```diff
@@ -27,9 +27,12 @@ export class SearchPage {
 
     async enableSearchFields(...fieldLabels: string[]) {
-        await this.page.getByRole('button', { name: 'Add Search Fields' }).click();
+        const addButton = this.page.getByRole('button', { name: 'Add Search Fields' });
+        // wait for button to be hydrated and visible
+        await expect(addButton).toBeVisible({ timeout: 10000 });
+        await addButton.click();
```
Ah I made this change while debugging the integration tests it is not required
```suggestion
        await this.page.getByRole('button', { name: 'Add Search Fields' }).click();
```

anna-parker committed:
Update integration-tests/tests/pages/search.page.ts

anna-parker commented on preprocessing/nextclade/tests/test_data/cchf_1/metadata.tsv:
```diff
@@ -0,0 +1,2 @@
+accessionVersion	accession	version	submissionId	isRevocation	submitter	groupName	groupId	submittedAtTimestamp	submittedDate	releasedAtTimestamp	releasedDate	dataUseTerms	dataUseTermsRestrictedUntil	dataUseTermsUrl	versionStatus	versionComment	pipelineVersion	ampliconPcrPrimerScheme	ampliconSize	anatomicalMaterial	anatomicalPart	authorAffiliations	authors	bioprojectAccession	biosampleAccession	bod ...
```
@fhennig what is the last metadata entry?
```
[{""fileId"":""7960e7ac-1282-4b39-8ca2-6c4a71e3fc30"",""name"":""LOC_000006N.embl"",""url"":""http://localhost:8079/files/get/LOC_000006N/1/annotations/LOC_000006N.embl""}]"
```
I didn't realize we needed this... I think it shouldn't be there - or am I missing sth about file upload?

fhennig replied:
Ah! Yes, not needed, I removed it. That was only there because I downloaded the metadata file, and then it was in there.

anna-parker reviewed (approved):
I reviewed the prepro code quite thoroughly a couple weeks ago so I now only reviewed the code that changed, focusing on the tests. Additionally, I submitted cchf (with 3 and 2 segments), revised and revoked the submitted sequence and confirmed the embl flat files were still available. 

I think the code it good to go but there are two comments to be resolved and I think because we had some issues with the prepro pipeline it would be good to do do a manual test where the pipeline version is updated. 

I would propose the following test before we merge:
1. the db is set to persistent and the embl flat files are all uploaded
2. the prepro pipeline is set to a higher version. We should then check that all embl flat files are updated and are all still available - I would suggest adding a note to the embl flat file header that this is the later prepro version to ensure the update was successful (should be reverted before merging) 
3. confirm all embl flat files are updated and available

fhennig committed:
remove unneccessary metadata field

fhennig commented on preprocessing/nextclade/src/loculus_preprocessing/embl.py:
```diff
@@ -0,0 +1,263 @@
+        gene_range = gene.get("range")
+        attributes = gene.get("attributes", {})
+        qualifiers = {
+            new_key: attributes[old_key]
```
Hmm, agreed that it doesn't look quite right, good catch!

Definitely good to look at the source, it seems a bit odd that this bit would be URL encoded, maybe it's not intended to come like this from nextclade?

anna-parker replied:
you can make an issue in nextclade then with this thread :-)

fhennig replied:
I'm marking this as resolved

anna-parker reviewed (changes requested):
sorry I thought the auth issues were fixed! but requesting changes until we figure out why they are still turning up

chaoran-chen commented:
Thanks, @anna-parker, for noticing the test failing! I identified the bug: #4580

fhennig commented:
I have a PR ready for the fix: #4581 - we can get that one merged, and rebase, and then things should be green 🤞

anna-parker reviewed (approved):
Awesome thanks for the additional tests! We can revert them before merging :-)

anna-parker committed:
feat(ingest): use nextclade in tests to ensure we can update nextclade without any issues

anna-parker committed:
feat(ingest): update nextclade

anna-parker committed:
feat(ingest): make code cleaner

anna-parker committed:
feat(ingest): improve typing

anna-parker committed:
feat(prepro): add test

anna-parker commented on preprocessing/nextclade/src/loculus_preprocessing/embl.py:
```diff
@@ -0,0 +1,263 @@
+
+def get_description(accession, version, db_name) -> str:
+    return (
+        f"Original sequence submitted to {db_name} with accession: {accession}, version: {version}"
```
I just realized - this description will have to be modified for all sequences that we ingest from the INSDC and we should probably also modify it if there is a gisaid accession listed

anna-parker commented on kubernetes/loculus/templates/loculus-preprocessing-config.yaml:
```diff
@@ -5,6 +5,9 @@
 {{- if $processingConfig.configFile }}
+{{- /* Use the enaDepositionConfig as the base config */}}
+{{- $enaDepositionConfig := dig "enaDeposition" "configFile" dict $organismConfig }}
+{{- $mergedConfigFile := merge $processingConfig.configFile $enaDepositionConfig }}
```
lets maybe rename this to not collide with the incoming multi-path changes

anna-parker committed:
feat(config): rename

anna-parker committed:
feat(prepro): improve file descriptions

anna-parker committed:
feat(prepro): clean up run function

theosanderson commented on preprocessing/nextclade/src/loculus_preprocessing/config.py:
```diff
@@ -15,9 +15,24 @@
+    molecule_type: str = "genomic RNA"
+    topology: str = "linear"
+    db_name: str = "Loculus"
+    # The backend host is the base API URL including the organism
```
This makes sense of the comment above but if so it's misnamed as "host"? But maybe this is separate to this PR

anna-parker replied:
https://github.com/loculus-project/loculus/issues/4816

anna-parker committed:
fix merge conflicts

anna-parker commented on preprocessing/nextclade/src/loculus_preprocessing/config.py:
```diff
@@ -54,6 +64,15 @@ class Config:
     accepted_dataset_matches: list[str] = dataclasses.field(default_factory=list)
+    create_embl_file: bool = False
+    scientific_name: str = "Orthonairovirus haemorrhagiae"
+    molecule_type: str = "genomic RNA"
```
technically both topology and molecule_type are enums

anna-parker commented on preprocessing/nextclade/src/loculus_preprocessing/embl.py:
```diff
@@ -0,0 +1,289 @@
+        if not sequence_str:
+            continue
+        reference = Reference()
+        segment = seq_name if multi_segment else None
```
```suggestion
        segment = seq_name if config.multi_segment else None
```

anna-parker committed:
Apply suggestions from code review

anna-parker committed:
use enums

anna-parker commented on kubernetes/loculus/values.yaml:
```diff
@@ -1204,6 +1206,7 @@ defaultOrganismConfig: &defaultOrganismConfig
         log_level: DEBUG
         genes: []
         batch_size: 100
+        create_embl_file: true
```
note for implementation in PPX: these fields also need to be added to the prepro config (they are in enaDepoistion config atm):
```
taxon_id: 186540
scientific_name: "Sudan ebolavirus"
molecule_type: "genomic RNA"
```
